packages/ListBox/src/components/Filter/Filter.js

```diff
@@ -0,0 +1,131 @@
+
+const propTypes = {
+  filter: PropTypes.func,
+  filterExcludeSelectedOptions: PropTypes.bool,
+  hasSearchIcon: PropTypes.bool,
+  noResultsMessage: PropTypes.string,
+  onChangeFilter: PropTypes.func,
+  onKeyDown: PropTypes.func,
+  placeholder: PropTypes.string,
+  renderFilter: PropTypes.func,
+  value: PropTypes.string,
+};
+
+const defaultProps = {
+  filter: null,
+  filterExcludeSelectedOptions: false,
+  hasSearchIcon: true,
+  onChangeFilter: null,
+  onKeyDown: null,
+  placeholder: "Filter...",
+  noResultsMessage: "Your search did not match any options.",
+  renderFilter: null,
+  value: null,
+};
+
+export default function Filter(props) {
+  const [state, dispatch] = useListBox();
+  const [textSearch, setTextSearch] = React.useState(props.value);
+  const applyFilterType = useListBox.types.applyFilter;
+
+  const handleChangeFilter = event => {
+    function hasNoResults(textSearchValue, filteredOptions) {
+      return textSearchValue && (filteredOptions && filteredOptions.length === 0);
+    }
+
+    const textSearchValue = event.target.value;
+
+    if (state.isDisabled) return;
+
+    if (props.filter) {
+      setTextSearch(textSearchValue);
+
+      const result = props.filter({ search: textSearchValue });
+      if (result && result.then) {
+        // detecting if it's a promise
+        result.then(response => {
+          const filteredOptions = response;
+          const noResultsFound = hasNoResults(textSearchValue, filteredOptions);
+          applyFilter(dispatch, applyFilterType)(filteredOptions, noResultsFound);
+        });
+      }
+      return;
+    }
+
+    if (props.onChangeFilter) {
+      props.onChangeFilter(event);
+    } else {
+      setTextSearch(textSearchValue);
+    }
+
+    const filteredOptions = filter({ props, state, textSearchValue });
+    const noResultsFound = hasNoResults(textSearchValue, filteredOptions);
+    applyFilter(dispatch, applyFilterType)(filteredOptions, noResultsFound);
+  };
+
+  const handleBlur = () => {
+    window.requestAnimationFrame(() => {
+      if (state.hasFooter) {
+        // if has footer the responsible of closing is the Footer no the onblur event
+        return;
+      }
+
+      if (document.activeElement !== state.refListBoxContainer.current) {
+        // this will reset the activeOption and close the Popover
+        dispatch({
+          type: useListBox.types.setActiveOption,
+          payload: {
+            activeOptionIndex: null,
+          },
+        });
+      }
+    });
+  };
+
+  const handleEffectValue = effects.handleEffectValue(props, applyFilter(dispatch, applyFilterType));
+  const handleEffectIsPopOverOpen = effects.handleEffectIsPopOverOpen(state, setTextSearch);
+  const handleEffectTextSearch = effects.handleEffectTextSearch(textSearch, applyFilter(dispatch, applyFilterType));
+  const handleEffectHasFilter = effects.handleEffectHasFilter(dispatch, useListBox.types.hasFilter);
```

Comment:
Hmm, I'm not so sure this layer of indirection is valuable, I think I'd prefer to see them inline below.